torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -198,6 +202,102 @@ bool is_numpy_scalar(PyObject* obj) {
           PyArray_IsScalar(obj, Floating));
 }
 
+at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
```

Comment:
I'm going to ask you to port the Python bindings here to use pybind11. The reason for this is that you are leaking memory in error conditions (I'll point this out shortly), and if you use actual RAII C++ objects, these problems will go away. It's much more scalable than having me manually run a "did you free the refcounted object on all logic paths" analysis on your code.